pinot-common/src/main/java/com/linkedin/pinot/common/partition/ReplicaGroupPartitionAssignmentGenerator.java

```diff
@@ -0,0 +1,121 @@
+    String path = ZKMetadataProvider.constructPropertyStorePathForInstancePartitions(tableNameWithType);
+    ZNRecord replicaGroupPartitionAssignment = _propertyStore.get(path, null, AccessOption.PERSISTENT);
+    ReplicaGroupPartitionAssignment partitionAssignment = null;
+    if (replicaGroupPartitionAssignment != null) {
+      partitionAssignment =
+          new ReplicaGroupPartitionAssignment(tableNameWithType, replicaGroupPartitionAssignment.getListFields());
+    }
+    return partitionAssignment;
+  }
+
+  /**
+   * Write the replica group partition assignment to property store
+   *
+   * @param partitionAssignment a replica group partition assignment
+   */
+  public void writeReplicaGroupPartitionAssignment(ReplicaGroupPartitionAssignment partitionAssignment) {
+    String tableNameWithType = partitionAssignment.getTableName();
+    ZNRecord znRecord = new ZNRecord(tableNameWithType);
+    znRecord.setListFields(partitionAssignment.getPartitionToInstances());
+    String path = ZKMetadataProvider.constructPropertyStorePathForInstancePartitions(tableNameWithType);
+    _propertyStore.set(path, znRecord, AccessOption.PERSISTENT);
+  }
+
+  /**
+   * Build the partition mapping table that maps a tuple of (partition number, replica group number) to a list of
+   * servers. Two important configurations are explained below.
+   *
+   * 'numInstancesPerPartition': this number decides the number of servers within a replica group.
+   * 'partitionColumn': this configuration decides whether to use the table or partition level replica groups.
+   *
+   * @param tableNameWithType: Name of table
+   * @param tableConfig: Configuration for table
+   * @return Replica group partition assignment from the given configuration
+   */
+  public ReplicaGroupPartitionAssignment buildReplicaGroupPartitionAssignment(String tableNameWithType,
+      TableConfig tableConfig, List<String> servers) {
+
+    // Fetch information required to build the mapping table from the table configuration.
+    ReplicaGroupStrategyConfig replicaGroupStrategyConfig = tableConfig.getValidationConfig().getReplicaGroupStrategyConfig();
+    String partitionColumn = replicaGroupStrategyConfig.getPartitionColumn();
+    int numInstancesPerPartition = replicaGroupStrategyConfig.getNumInstancesPerPartition();
+
+    // If we do not have the partition column configuration, we assume to use the table level replica groups,
+    // which is equivalent to have the same partition number for all segments (i.e. 1 partition).
+    int numPartitions = 1;
+    if (partitionColumn != null) {
+      numPartitions = tableConfig.getIndexingConfig().getSegmentPartitionConfig().getNumPartitions(partitionColumn);
+    }
+    int numReplicas = tableConfig.getValidationConfig().getReplicationNumber();
+    int numServers = servers.size();
+
+    // Enforcing disjoint server sets for each replica group.
+    if (numInstancesPerPartition * numReplicas > numServers) {
```

Comment:
numPartitions and numInstancesPerPartition terminology is a little confusing. 
numPartitions is referring to the partitions on a column right? so 0_* 1_* means 0, 1 are the partitions i.e.numPartitions=2. 
numInstancesPerPartition seems to refer to 0_0 0_1 .. 0_n as separate partitions. So it is actually referring to numInstancesPerReplica

Can we change one of them? I see that numInstancesPerPartition is already in the tableConfig. So maybe we can rename numPartitions to something like numColumnPartitions

Reply:
`numInstancesPerReplicaGroup` will be better name but as you said, it's already in table config. I changed the name for `numPartitions`